composer/utils/object_store/sftp_object_store.py

```
@@ -0,0 +1,84 @@
+
+    Args:
+        host (str): The server to connect to
+        port (int): The server port to connect to
+        username (str): The username to authenticate
+        key_file_path (str): The filename of private key
+        cwd (Optional[str]): The directory to navigate to upon creating the SSH connection.
+    """
+
+    def __init__(self, host: str, port: int, username: str, key_file_path: str, cwd: Optional[str] = None):
+        self.host = host
+        self.port = port
+        self.username = username
+        self.key_file_path = key_file_path
+        self.cwd = cwd
+
+        try:
+            from paramiko import SSHClient
+        except ImportError as e:
+            raise MissingConditionalImportError(extra_deps_group='streaming', conda_package='paramiko') from e
+        self.ssh_client = SSHClient()
+
+        try:
+            self.ssh_client.load_system_host_keys(key_file_path)
+        except IOError:
+            raise Exception('Host keys could not be read from {key_file_path}.')
+
+        self.ssh_client.connect(host, port, username)
+        if cwd is not None:
+            self.ssh_client.exec_command(f'mkdir -p {cwd}')
+            self.ssh_client.exec_command(f'cd {cwd}')
+
+        self.sftp_client = self.ssh_client.open_sftp()
+
+    def close(self):
+        self.ssh_client.close()
+
+    def get_uri(self, object_name: str) -> str:
+        return f'sftp://{self.host}:{self.port}/{object_name}'
+
+    def get_object_size(self, object_name: str) -> int:
+        return self.sftp_client.stat(object_name).st_size
+
+    def upload_object(self, file_path: str, object_name: str):
+        tmp_path = object_name + f'.{uuid.uuid4()}.tmp'
+
+        dirname = os.path.dirname(object_name)
+        self.ssh_client.exec_command(f'mkdir -p {dirname}')
+
+        try:
+            self.sftp_client.put(file_path, tmp_path, confirm=True)
+        except IOError:
+            raise ObjectStoreTransientError
```

Are all IO errors always transient errors? And are there other types of non-io transient exceptions that could be raised? In addition, can we re-raise the exception?

```suggestion
            raise ObjectStoreTransientError from e
```

---

Resolved